File: answers/leetcode/108. Convert Sorted Array to Binary Search Tree/Solution.cpp
// 2014-04-17
//@result 32 / 32 test cases passed. Status: Accepted Runtime: 16 ms Submitted: 0 minutes ago You are here! Your runtime beats 68.53% of cpp submissions.

class Solution {
public:
	TreeNode *sortedArrayToBST(vector<int> &num) {
		return sortedArrayToBST(num, 0, num.size());
	}
	TreeNode *sortedArrayToBST(const vector<int> &num, int left, int right)
	{
		if (left >= right)
		{
			return NULL;
		}
		int mid = (left + right) / 2;
		TreeNode *root(new TreeNode(num[mid]));
		root->left = sortedArrayToBST(num, left, mid);
		root->right = sortedArrayToBST(num, mid + 1, right);
		return root;
	}
};

File: answers/leetcode/108. Convert Sorted Array to Binary Search Tree/Solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        if nums == None or len(nums) == 0:
            return None
        else:
            return self.sort(0, len(nums) - 1, nums)

    def sort(self, start: int, end: int, nums: List[int]) -> Optional[TreeNode]:
        if start > end:
            return None
        elif start == end:
            return TreeNode(nums[start])
        else:
            middle: int = (start + end) // 2
            root: TreeNode = TreeNode(nums[middle])
            root.left = self.sort(start, middle - 1, nums)
            root.right = self.sort(middle + 1, end, nums)
            return root
